### ISSUE TITLE
ChildRouter in ViewPager, Not working

### ISSUE THREAD

**Issue by Rawa** [bluelinelabs/Conductor#426]
Hello!

I'm building trying to build a app which has a `TabLayout` with `ViewPager`, where in one of the tabs a different controller will push a `controller` to a `ChildRouter` depending on the state of the app. Basically I got the following scheme in my app:
`MainActivity` - Setup router
`MainController` -> RootController which manages the TabLayout and ViewPager
`TabController` -> Abstract class that helps MainController build the tabs.
`SimpleTabController` and `ForwardTabController` -> Tabs that extends the TabController class.
`FirstController` and `SecondController` -> Two samples that are to be pushed onto the `ForwardTabController` depending on the state of the app.
`BaseController` -> Common stuff for controllers.

I have a proof of concept of the problem I'm experiencing here:
https://github.com/Rawa/conductorVIewPager 

My problem occurs with the recycling of the ForwardTabController. When I go from the furthest left tab, ForwardTabView controller to the tab furthest right the `ViewPager` will detatch and destroy the first controller (intended) and then when i switch back to the first tab `onDestroy` is called (again?!) on the attach of the new controller I get the following exception:

`03-28 11:31:27.694 20419-20419/com.technocreatives.pansaraugmented E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.technocreatives.pansaraugmented, PID: 20419
    java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.ViewParent android.view.View.getParent()' on a null object reference
        at com.bluelinelabs.conductor.Controller.attach(Controller.java:879)
        at com.bluelinelabs.conductor.Controller.access$500(Controller.java:44)
        at com.bluelinelabs.conductor.Controller$7.onAttached(Controller.java:1005)
        at com.bluelinelabs.conductor.internal.ViewAttachHandler.reportAttached(ViewAttachHandler.java:87)
        at com.bluelinelabs.conductor.internal.ViewAttachHandler.access$100(ViewAttachHandler.java:7)
        at com.bluelinelabs.conductor.internal.ViewAttachHandler$1.onAttached(ViewAttachHandler.java:47)
        at com.bluelinelabs.conductor.internal.ViewAttachHandler$2.onViewAttachedToWindow(ViewAttachHandler.java:126)
        at android.view.View.dispatchAttachedToWindow(View.java:17456)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3333)
        at android.view.ViewGroup.addViewInner(ViewGroup.java:4977)
        at android.view.ViewGroup.addView(ViewGroup.java:4768)
        at android.support.v4.view.ViewPager.addView(ViewPager.java:1479)
        at android.view.ViewGroup.addView(ViewGroup.java:4708)
        at android.view.ViewGroup.addView(ViewGroup.java:4681)
        at com.bluelinelabs.conductor.changehandler.SimpleSwapChangeHandler.performChange(SimpleSwapChangeHandler.java:73)
        at com.bluelinelabs.conductor.ControllerChangeHandler.executeChange(ControllerChangeHandler.java:204)
        at com.bluelinelabs.conductor.ControllerChangeHandler.executeChange(ControllerChangeHandler.java:154)
        at com.bluelinelabs.conductor.Router.performControllerChange(Router.java:774)
        at com.bluelinelabs.conductor.Router.performControllerChange(Router.java:744)
        at com.bluelinelabs.conductor.Router.rebindIfNeeded(Router.java:501)
        at com.bluelinelabs.conductor.support.RouterPagerAdapter.instantiateItem(RouterPagerAdapter.java:77)
        at android.support.v4.view.ViewPager.addNewItem(ViewPager.java:1004)
        at android.support.v4.view.ViewPager.populate(ViewPager.java:1152)
        at android.support.v4.view.ViewPager.setCurrentItemInternal(ViewPager.java:663)
        at android.support.v4.view.ViewPager.setCurrentItemInternal(ViewPager.java:625)
        at android.support.v4.view.ViewPager.setCurrentItem(ViewPager.java:606)
        at android.support.design.widget.TabLayout$ViewPagerOnTabSelectedListener.onTabSelected(TabLayout.java:2170)
        at android.support.design.widget.TabLayout.dispatchTabSelected(TabLayout.java:1165)
        at android.support.design.widget.TabLayout.selectTab(TabLayout.java:1158)
        at android.support.design.widget.TabLayout.selectTab(TabLayout.java:1128)
        at android.support.design.widget.TabLayout$Tab.select(TabLayout.java:1427)
        at android.support.design.widget.TabLayout$TabView.performClick(TabLayout.java:1537)
        at android.view.View$PerformClick.run(View.java:24770)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)`

I've tried several combinations of pushing the controller, setting the root controller, removeChildRouter onDestroy. Without managing to get it to work.

Thanks

**Comment by Rawa** [comment]
Did some more research and tried using the deprecated ControllerPagerAdapter instead with saveState set to false everything works fine. It seems that saving and restoring the state when having a ChildRouter could be the cause of this?

**Comment by Rawa** [comment]
Oh my, I finally solved my issue. So after doing some researching and having some luck i found the cause of the issue.

So what I tried doing is modifying the stack of a childRouter in onAttach or in onPostAttach. However, it seems like onPostAttach does not imply that the child controllers etc have been attached.

Originally I thought the issue was related to the saving and restoring of state which is not fully correct. What happens is that I popToRoot in onAttach/onPostAttach, which causes the restoring of the childRouter to fail. Hence why the backstack didn't correctly inflate. I solved the issue in my latest commit, one are not simply not allowed to change the child router until onChangeEnded handler has been called.
